### ISSUE TITLE
Can't build for GPU on MacOS system

### ISSUE THREAD

**Issue by mehranshakeri** [DLTcollab/dcurl#45]
I'm trying to build `make BUILD_GPU=1 check` and it fails with the following error
```
grep: /proc/cpuinfo: No such file or directory
grep: /proc/cpuinfo: No such file or directory
  CC	build/curl.o
  CC	build/constants.o
  CC	build/trinary.o
  CC	build/dcurl.o
  CC	build/pow_c.o
  CC	build/pow_cl.o
src/pow_cl.c:9:10: fatal error: 'CL/cl.h' file not found
#include <CL/cl.h>
         ^~~~~~~~~
1 error generated.
make: *** [build/pow_cl.o] Error 1
```

I'm using OS X (Sierra 10.13.4) (Radeon Pro 460 4096 MB). 
I read the instructions and I actually don't know what to change in `clcontext.c` but in `opencl.mk` I set 
```
OPENCL_LIB ?= /System/Library/Frameworks/OpenCL.framework/OpenCL
```
and I'm not sure about the value for `OPENCL_PATH`.

I tried to google around but nothing helpful came across. 
Could you please specify exactly which parameters in `clcontext.c` should be set and give an example for `OPENCL_LIB` and `OPENCL_PATH`? Or give us some hints how to find them?

Thanks in advance. 
ps: I could build it for CPU successfully and I used the library as well. I just had to change `.so` to `.dylib` in `Makefile` to be able to use it in NodeJS library `iota.lib.js`

**Comment by furuame** [comment]
Hi, thanks for your report. Sadly I don't have a Macbook on my side and debug for it.

Have you ever tried the following to replace original code?
```
#if defined(__APPLE__) || defined(__MACOSX)
#include <OpenCL/cl.h>
#else
#include <CL/cl.h>
#endif
```

If this patch works, welcome to issue a pull request back :) 
`OPENCL_PATH` can be ignored, which is not used anymore now.

For parameters in `clcontext.c`, I think you can just skip it right now if you build dcurl on MacBook.
I'm not really sure how to find OpenCL library on MacOS but you can try the above solution first.

**Comment by mehranshakeri** [comment]
Thank you for quick reply.
With following `opencl.mk` 
```
# FIXME: perform sanity checks before compilation
OPENCL_LIB ?= /System/Library/Frameworks/OpenCL.framework

CFLAGS += \
        -DENABLE_OPENCL \
        -I$(OPENCL_PATH)/include
LDFLAGS += -L$(OPENCL_LIB) -lOpenCL
```

I get this error
```
grep: /proc/cpuinfo: No such file or directory
  CC	build/test-trinary.o
  LD	build/test-trinary
ld: library not found for -lOpenCL
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [build/test-trinary] Error 1
rm build/test-trinary.o
```

And if I remove `-lOpenCL` from `LDFLAGS` in `opencl.mk` file, I get this error:
```
grep: /proc/cpuinfo: No such file or directory
  CC	build/test-trinary.o
  LD	build/test-trinary
Undefined symbols for architecture x86_64:
  "_clBuildProgram", referenced from:
      _init_clcontext in clcontext.o
  "_clCreateBuffer", referenced from:
      _init_cl_buffer in clcontext.o
  "_clCreateCommandQueue", referenced from:
      _init_clcontext in clcontext.o
  "_clCreateContext", referenced from:
      _init_clcontext in clcontext.o
  "_clCreateKernel", referenced from:
      _init_cl_kernel in clcontext.o
  "_clCreateProgramWithSource", referenced from:
      _init_clcontext in clcontext.o
  "_clEnqueueNDRangeKernel", referenced from:
      _PowCL in pow_cl.o
  "_clEnqueueReadBuffer", referenced from:
      _PowCL in pow_cl.o
  "_clEnqueueWriteBuffer", referenced from:
      _PowCL in pow_cl.o
  "_clGetDeviceIDs", referenced from:
      _init_clcontext in clcontext.o
  "_clGetDeviceInfo", referenced from:
      _init_clcontext in clcontext.o
  "_clGetPlatformIDs", referenced from:
      _init_clcontext in clcontext.o
  "_clReleaseEvent", referenced from:
      _PowCL in pow_cl.o
  "_clSetKernelArg", referenced from:
      _init_cl_buffer in clcontext.o
  "_clWaitForEvents", referenced from:
      _PowCL in pow_cl.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [build/test-trinary] Error 1
rm build/test-trinary.o
```

**Comment by furuame** [comment]
I think you could try this [solution](https://devtalk.nvidia.com/default/topic/484331/cannot-find-lopencl-on-mac-os-x-opencl-is-not-working-on-mac-os-as-it-cannot-find-the-library/)

Replace `-lOpenCL` with `-framework OpenCL` in Makefile

**Comment by mehranshakeri** [comment]
Yes, I was reading the same post :) 
Thank you again. I could build it. I will create a PR

**Comment by mehranshakeri** [comment]
Ok, There is another issue. :D 
I tried to use `libdcurl.dylib` from build dir in my code before creating the pull request and it fails with this error: 

```
/iotaprj/node_modules/ffi/lib/dynamic_library.js:112
throw new Error('Dynamic Symbol Retrieval Error: ' + this.error())
^

Error: Dynamic Symbol Retrieval Error: dlsym(0x10221c310, ccurl_pow): symbol not found
at DynamicLibrary.get (/iotaprj/node_modules/ffi/lib/dynamic_library.js:112:11)
at /iotaprj/node_modules/ffi/lib/library.js:50:19
at Array.forEach (<anonymous>)
at Object.Library (/iotaprj/node_modules/ffi/lib/library.js:47:28)
at ccurl (/iotaprj/node_modules/iota.lib.js/lib/api/api.js:2098:26)
at api.attachToTangle (/iotaprj/node_modules/iota.lib.js/lib/api/api.js:101:17)
at /iotaprj/node_modules/iota.lib.js/lib/api/api.js:552:14
at makeRequest.prepareResult (/iotaprj/node_modules/iota.lib.js/lib/utils/makeRequest.js:315:12)
at exports.XMLHttpRequest.request.onreadystatechange (/iotaprj/node_modules/iota.lib.js/lib/utils/makeRequest.js:71:25)
at exports.XMLHttpRequest.dispatchEvent (/iotaprj/node_modules/xmlhttprequest/lib/XMLHttpRequest.js:591:25)

```

But `make BUILD_GPU=1 check` passed successfully:

```
grep: /proc/cpuinfo: No such file or directory
grep: /proc/cpuinfo: No such file or directory
  CC	build/curl.o
  CC	build/constants.o
  CC	build/trinary.o
  CC	build/dcurl.o
  CC	build/pow_c.o
  CC	build/pow_cl.o
  CC	build/clcontext.o
  CC	build/test-trinary.o
  LD	build/test-trinary
*** Validating build/test-trinary ***
	[ Verified ]
  CC	build/test-curl.o
  LD	build/test-curl
*** Validating build/test-curl ***
	[ Verified ]
  CC	build/test-pow_c.o
  LD	build/test-pow_c
*** Validating build/test-pow_c ***
	[ Verified ]
  LD	build/libdcurl.dylib
Makefile:135: "skip build/test-multi_pow_cpu because PyIOTA is not installed."
*** Validating build/test-multi_pow_cpu ***
	[ Verified ]
  CC	build/test-pow_cl.o
  LD	build/test-pow_cl
*** Validating build/test-pow_cl ***
	[ Verified ]
Makefile:135: "skip build/test-multi_pow_gpu because PyIOTA is not installed."
*** Validating build/test-multi_pow_gpu ***
	[ Verified ]
rm build/test-trinary.o build/test-curl.o build/test-pow_cl.o build/test-pow_c.o
```

I am also able to do transactions with CPU only build.

Should I install PyIOTA?

**Comment by mehranshakeri** [comment]
Ok, I had to build it with extra flag :)
`make BUILD_GPU=1 BUILD_COMPAT=1 check`

-ps: I keep my previous comments in case if someone else faces the same mistake I've made :D

**Comment by furuame** [comment]
But I am curious why `libdcurl.dylib`, which build only for CPU, works well on your code

**Comment by furuame** [comment]
PyIOTA is used by the test program, which verifies the result of dcurl.

I think it had better be done before you use dcurl.

**Comment by mehranshakeri** [comment]
I did some tests. I can build the lib and use it, but despite the fact that I build it only for GPU, it uses CPU for PoW. (`make BUILD_GPU=1 BUILD_COMPAT=1 check`)
I tried to use `dcurl_init(0, 2)` which I guess it means running only with GPU but still it uses only CPU. Could it be because I'm running AMD graphic card?

**Comment by mehranshakeri** [comment]
What is PyIOTA and where can I get it? I couldn't find it in Google :))

**Comment by furuame** [comment]
I don't think it is normal... May I ask how do you observe that dcurl only use CPU?

PyIOTA is the package name of iota.lib.py, you can find the installation instructions in its github. Noted that test program is run by Python3.

**Comment by mehranshakeri** [comment]
When I built it only for CPU, I ran the transfer and it took less than 3 min to finish it. With new build it's the same time. I also use activity monitor, observing cpu load and I see all 8 cores will be 100% working during the PoW. I actually didn't monitor GPU activity. I assume if the load was on GPU, there shouldn't be any high load on CPU.

I try PyIOTA and see if it can help :)

**Comment by mehranshakeri** [comment]
It seems it's renamed to `pyota`. I installed it and all tests passed with no error or warning. But still high load on CPU.

**Comment by furuame** [comment]
Could you re-run `./build/pow_cl` again and observe CPU activity again?

**Comment by mehranshakeri** [comment]
There is no such a file. This is the list of files in build dir:
```
clcontext.o		constants.o.d		libdcurl.dylib		pow_cl.o.d		test-multi_pow_gpu	test-trinary
clcontext.o.d		curl.o			local.mk		test-ccurl-multi_pow	test-pow_c		test-trinary.o.d
compat-ccurl.o		curl.o.d		pow_c.o			test-curl		test-pow_c.o.d		trinary.o
compat-ccurl.o.d	dcurl.o			pow_c.o.d		test-curl.o.d		test-pow_cl		trinary.o.d
constants.o		dcurl.o.d		pow_cl.o		test-multi_pow_cpu	test-pow_cl.o.d
```

**Comment by furuame** [comment]
Sorry, it's `./build/test-pow_cl`

**Comment by mehranshakeri** [comment]
No CPU load at all.
It also doesn't print any output. Is it expected?

**Comment by furuame** [comment]
Yes, it is the expected result. And this program `./build/test-pow_cl` just uses GPU to compute do PoW.

You can modify [`./tests/test-multi_pow_gpu.py`](https://github.com/chenwei-tw/dcurl/blob/1fb527f4e6dd91c70d388837e97234be9105f268/tests/test-multi_pow_gpu.py#L83) Line 83 with (0, 1), which means dcurl would be initialized only for GPU.

Then please execute `$ python3 ./tests/test-multi_pow_gpu.py` again and observe cpu & gpu activity.

**Comment by mehranshakeri** [comment]
Using XRG I could see that this test is using my first GPU which is intel and is almost as slow as my CPU :D 
I've already disabled intel VGA for my OsX. Are you accessing the GPUs directly? If yes then how can I select second GPU which is AMD and faster?

**Comment by furuame** [comment]
Okay, It seems that configuration of OpenCL on MacBook is not quite similar to the normal desktop.

I think you can try to modify Line35@[`./src/clcontext.c`](https://github.com/chenwei-tw/dcurl/blob/1fb527f4e6dd91c70d388837e97234be9105f268/src/clcontext.c#L35) with the following:
```
clCreateContext(NULL, 1, &ctx->device[1], NULL, NULL, &errno)
```
which means using 2th device (GPU) at this platform

You can try it and observe the activity again.

**Comment by mehranshakeri** [comment]
```
    ctx->context =
        (cl_context) clCreateContext(NULL, 1, &ctx->device[1], NULL, NULL, &errno);
```

Build failed:

```
grep: /proc/cpuinfo: No such file or directory
grep: /proc/cpuinfo: No such file or directory
  CC	build/curl.o
  CC	build/constants.o
  CC	build/trinary.o
  CC	build/dcurl.o
  CC	build/pow_c.o
  CC	build/pow_cl.o
  CC	build/clcontext.o
src/clcontext.c:36:59: error: subscript of pointer to incomplete type 'struct _cl_device_id'
        (cl_context) clCreateContext(NULL, 1, &ctx->device[1], NULL, NULL, &errno);
                                               ~~~~~~~~~~~^
/System/Library/Frameworks/OpenCL.framework/Headers/cl.h:40:16: note: forward declaration of 'struct _cl_device_id'
typedef struct _cl_device_id *      cl_device_id;
               ^
1 error generated.
make: *** [build/clcontext.o] Error 1

```

**Comment by furuame** [comment]
I will deal with it and try to offer some suggestions how to adjust OpenCL configurations in these days.

By the way, after doing some experiments, I found SSE implementation, which is default implementation for CPU, is faster than GPU. I think you can use CPU only dcurl if you need :)

**Comment by mehranshakeri** [comment]
Thanks a lot for following the issue :D

**Comment by furuame** [comment]
Hi, @mehranshakeri 
Can you try the branch `opencl-second-device` again and observe gpu activity again?
(Remember to modify the `OPENCL_LIB` and linking file)

Let me brief the OpenCL configuration of dcurl:
* For convenience, dcurl only use the `first platform` and the `first device` at this time because I still have no idea how to let users configure each device of their machines if I want to enable each device to do different PoW tasks. (ccurl simply let every device on all platforms do the same PoW task)
* Patch in the `opencl-second-device` can use the `second device` on the `first platform`, you can see that how devices to be set in `src/clcontext.c` trivially from [here](https://github.com/chenwei-tw/dcurl/commit/67b965bf5568ef1f227d6b0812653049281c8583)
* I have verified this branch on my machine, which has two GPU cards, and only the second GPU does the PoW.
* I think your MacBook Pro may have `two` devices on the first platform, so you can try the patch to see whether it works.
* There is also a possible situation that GPU in the MacBook Pro is the `first device` on the `second platform`, I really not sure but I think you can try it.

**Comment by mehranshakeri** [comment]
Hi @chenwei-tw 
Thanks again for your efforts. I tested this branch and running multi gpu test shows the load on the correct GPU. So the patch is working :D 
To make it work with `iota.lib.js` I modified `compat-ccurl.c` line 20 to `dcurl_init(0, 1);` and I could see it's working fine.
Now the question is, is there any way to benchmark GPU vs CPU? Since finding nonce is random, running iota transfer is not a good measure :D 
Since you mentioned SSE, I'm interested to know should I stick with CPU or not.

**Comment by chenwei-tw** [comment]
You can use `tests/test-multi_pow_gpu.py` to measure and I have offered many trytes in `tests/trytes.txt`.
(Of course, you need to modify `NUM_TRYTES` and `test_set` in this test script)

Noted that build the dcurl with GPU support before using `tests/test-multi_pow_gpu.py`
You can modify `test_set` to test any configuration (CPU only / GPU only / CPU-GPU mixed).

Also you can reference @gagathos's [work](https://github.com/gagathos/iota-pow-benchmark), which can generate transaction trytes and pass them to PoW server. You can find some instructions in issue #42.

Issue fixed and close #45

**Comment by furuame** [comment]
You can use `tests/test-multi_pow_gpu.py` to measure and I have offered many trytes in `tests/trytes.txt`.
(Of course, you need to modify `NUM_TRYTES` and `test_set` in this test script)

Noted that build the dcurl with GPU support before using `tests/test-multi_pow_gpu.py`
You can modify `test_set` to test any configuration (CPU only / GPU only / CPU-GPU mixed).

Also you can reference @gagathos's [work](https://github.com/gagathos/iota-pow-benchmark), which can generate transaction trytes and pass them to PoW server. You can find some instructions in issue #42.

Issue fixed and close #45
